src/math/dd/dd_bdd.cpp

```diff
@@ -896,87 +911,54 @@ namespace dd {
+            return find_result::empty;
         bool is_unique = true;
-        unsigned num_vars = 0;
+        unsigned var_idx = bits.size();
         while (!is_true(b)) {
-            ++num_vars;
+            VERIFY(var_idx-- > 0);
+            SASSERT(var(b) <= bits[var_idx]);
+            while (var(b) < bits[var_idx]) {
+                is_unique = false;
+                VERIFY(var_idx-- > 0);
+            }
             if (!is_false(lo(b)) && !is_false(hi(b)))
                 is_unique = false;
             if (is_false(lo(b))) {
-                val += rational::power_of_two(var(b));
+                SASSERT(var(b) == bits[var_idx]);
+                val += rational::power_of_two(var_idx);
                 b = hi(b);
             }
             else
                 b = lo(b);
         }
-        is_unique &= (num_vars == w);
+        if (var_idx > 0)
+            is_unique = false;
 
-        return is_unique ? find_int_t::singleton : find_int_t::multiple;
+        return is_unique ? find_result::singleton : find_result::multiple;
     }
 
-    bdd bdd_manager::mk_affine(rational const& a, rational const& b, unsigned w) {
-        if (a.is_zero())
-            return b.is_zero() ? mk_true() : mk_false();
-        // a*x + b == 0 (mod 2^w)
-        unsigned const rank_a = a.trailing_zeros();
-        unsigned const rank_b = b.is_zero() ? w : b.trailing_zeros();
-        // We have a', b' odd such that:
-        // 2^rank(a) * a'* x  +  2^rank(b) * b'  ==  0  (mod 2^w)
-        if (rank_a > rank_b) {
-            // <=> 2^(rank(a)-rank(b)) * a' * x  +  b'  ==  0  (mod 2^(w-rank(b)))
-            // LHS is always odd => equation cannot be true
-            return mk_false();
-        }
-        else if (b.is_zero()) {
-            // this is just a specialization of the else-branch below
-            return mk_int(rational::zero(), w - rank_a);
-        }
-        else {
-            unsigned const j = w - rank_a;
-            // Let b'' := 2^(rank(b)-rank(a)) * b', then we have:
-            // <=> a' * x  +  b''  ==  0  (mod 2^j)
-            // <=> x  ==  -b'' * inverse_j(a')  (mod 2^j)
-            // Now the question is, for what x in Z_{2^w} does this hold?
-            // Answer: for all x where the lower j bits are the same as in the RHS of the last equation,
-            // so we just fix those bits and leave the others unconstrained
-            // (which we can do simply by encoding the RHS as a j-width integer).
-            rational const pow2_rank_a = rational::power_of_two(rank_a);
-            rational const aa = a / pow2_rank_a;  // a'
-            rational const bb = b / pow2_rank_a;  // b''
-            rational inv_aa;
-            VERIFY(aa.mult_inverse(j, inv_aa));
-            rational const cc = mod(inv_aa * -bb, rational::power_of_two(j));
-            return mk_int(cc, j);
-        }
-   }
-
 
     bdd bdd_manager::mk_eq(bddv const& a, bddv const& b) {
+        SASSERT(a.size() == b.size());
```

Comment:
we might consider to run the loop from below instead of above to follow common case variable ordering (in the case this leads to fewer overall operations to build bdd). Maybe easy to test by implementing two versions and counting statistics in a test.

Reply:
Yes, counting from below seems to be slightly better. To compare, I printed `m_node_table.size()` as an approximation for the number of operations (idea: counting the "real" operations that don't just look up from the cache).